flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java

```diff
@@ -87,18 +88,19 @@
 
     private static final Logger LOG = LoggerFactory.getLogger(ZooKeeperStateHandleStore.class);
 
-    @VisibleForTesting
-    static final Set<Class<? extends KeeperException>> PRE_COMMIT_EXCEPTIONS =
-            newHashSet(
-                    KeeperException.NodeExistsException.class,
-                    KeeperException.BadArgumentsException.class,
-                    KeeperException.NoNodeException.class,
-                    KeeperException.NoAuthException.class,
-                    KeeperException.BadVersionException.class,
-                    KeeperException.AuthFailedException.class,
-                    KeeperException.InvalidACLException.class,
-                    KeeperException.SessionMovedException.class,
-                    KeeperException.NotReadOnlyException.class);
+    /** Pre-commit exceptions that don't imply data inconsistency. */
+    private static final Set<Class<? extends KeeperException>> SAFE_PRE_COMMIT_EXCEPTIONS =
+            new HashSet<>(
+                    Arrays.asList(
+                            KeeperException.NodeExistsException.class,
+                            KeeperException.BadArgumentsException.class,
+                            KeeperException.NoNodeException.class,
+                            KeeperException.NoAuthException.class,
+                            KeeperException.BadVersionException.class,
+                            KeeperException.AuthFailedException.class,
+                            KeeperException.InvalidACLException.class,
+                            KeeperException.SessionMovedException.class,
+                            KeeperException.NotReadOnlyException.class));
```

Comment:
Should we double check that any of these exceptions is safe in the sense that if we see it, then no metadata has been written (for `NodeExistsException` this does not hold true)? Maybe it is possible that we write something, then there is a connection loss and then a session moved. Can this happen? If yes, then we shouldn't consider `SessionMovedException`, for example, as safe.

Reply:
I'd say in very rare conditions it could happen (if connection loss happens right before retries are exhausted). We default to 3 retries with exponential backoff.

We could also argue that this could happen `AuthFailed` / `InvalidACLException` when there is some manual operation on the broker side. Eg. sys-admin changes the ACL

Description of `SessionMovedException`
```
SESSIONMOVED
Session moved to another server, so operation is ignored
```

From `RetryLoop`:
```
public static boolean shouldRetry(int rc) {
    return rc == Code.CONNECTIONLOSS.intValue() || rc == Code.OPERATIONTIMEOUT.intValue() || rc == Code.SESSIONMOVED.intValue() || rc == Code.SESSIONEXPIRED.intValue() || rc == -13;
}
```

Reply:
I'm just desperate here... I'd say except `NoNodeException` and `BadArgumentsException` all of them are "potentially" unsafe if preceded by connection loss.

I'd be in favor of removing this list completely as it's really hard to reason about.

I'd even suggest reverting https://issues.apache.org/jira/browse/FLINK-22494 and coming up with a different solution. (discarding checkpoint & validating metadata before recovery)

Reply:
One more suggestion, let's merge it as is and we should follow-up in different PR. It's still a big improvement from the original state and we'd need the patch for the transaction code path anyway.

Reply:
Agreed.